lib/music/note.rb

```diff
@@ -244,20 +245,29 @@ def chord(description)
     end
 
     class << self
-      def parse_note_string(note_string, assumed_octave = nil)
+      def parse_note_string(note_string, assumed_octave = nil, convert_enharmonics = nil)
         match = note_string.match(/^([A-Ga-g])([#b]?)([0-8]?)$/)
 
         fail ArgumentError, "Did not recognize note string: #{note_string}" if !match
         fail ArgumentError, 'No octave found or specified' if match[3].empty? && assumed_octave.nil?
         fail ArgumentError if match[3].to_i > 8 || (assumed_octave && !(0..8).include?(assumed_octave))
 
         octave = match[3].empty? ? assumed_octave : match[3]
-        [match[1].upcase, match[2] == '' ? nil : match[2], octave.to_i]
+        note_parts = [match[1].upcase, match[2] == '' ? nil : match[2], octave.to_i]
+        convert_enharmonics ? convert_enharmonics(*note_parts) : note_parts
+      end
+
+      def convert_enharmonics(note, accidental, octave)
+        full_note = [note, accidental].join
+        enharmonic_equivalent = ENHARMONICS[full_note]
+        return [note, accidental, octave] unless enharmonic_equivalent
+
+        [enharmonic_equivalent, nil, octave]
       end
```

Comment:
This feels kind of "patched on top".  The `NOTES` constant already includes some enharmonics (though I didn't know what it was called at the time 😆 ), but I think maybe it's just incomplete.  So just changing it to this might fix it:

```ruby
    NOTES = ['C/B#', 'C#/Db', 'D', 'D#/Eb', 'E/Fb', 'F/E#', 'F#/Gb', 'G', 'G#/Ab', 'A', 'A#/Bb', 'B/Cb']
```

Looking at that, maybe there's something that should be done for D and G 😄 

Looking at this now, I'm also wondering if my solution was a bit clunky.  Maybe it could be this and we wouldn't need to use regex:

```ruby
    NOTES = [['C', 'B#'], ['C#', 'Db'], ['D'], ['D#', 'Eb'], ['E', 'Fb'], ['F', 'E#'], ['F#', 'Gb'], ['G'], ['G#', 'Ab'], ['A'], ['A#', 'Bb'], ['B', 'Cb']]
```

Reply:
Yes I agree, I was a bit reluctant to touch the `NOTES` constant as it's used throughout so tried to constrain my changes just to parsing the note strings.

But I think you're right, I'll have a go at this approach and getting rid of the regex solution this weekend and let you know how I get on!

Thank you for such a detailed review! 😸

Reply:
👍  Great, yeah, feel free to change things if it makes things better! 🎉

Reply:
I have implemented this change and all the tests pass! But... I don't think it is quite right as the whole `give_flat` functionality is now slightly wrong.

For example [here](https://github.com/stufro/music/blob/handle_enharmonics/lib/music/note.rb#L309-L313), we decide whether to give the first or last note (G# or Ab) based on whether we want the sharp or flat spelling of a note.

This works for most notes e.g. `['G#', 'Ab']` but with these special enharmonics it doesn't e.g. `['C', 'B#']`. This would mean if you had a C5 note and called `.letter(give_flat: true)` it would return `B#` instead of `C`. If you switch the order around in the array e.g. `['B#', 'C']` you have a similar problem but in an opposite way (hope that makes sense!)

This makes me feel that handling these special harmonics should be a separate case and not built into the `NOTES` constant. What do you reckon?

Reply:
Ok, so I have various thoughts/ideas 😆 

Firstly, this makes me think that maybe `give_flat` was a bad choice for an argument (or at least an argument name).  It could have been `prefer_flat`.  But maybe instead there could have been a `prefer` option which could be either `:sharp`, `:flat`, or `:base`.  But then if you give `base` and the two options for the note are flat and sharp, then what do you choose? 😄

But I guess it depends on what's important for the option.  I honestly am not sure what people would want when using the gem.  And apparently I didn't write any tests for `give_flat` 😅  But I think that if this code works that maybe we should figure out what works for this particular change and we could work in another PR on this if you're interested.  It might involve introducing a breaking change and thus a new major version, so maybe good to keep it separate.

I think maybe the simplest approach for right now is to allow `give_flat: true` to return the flat when possible and otherwise return the base/sharp note.  With `give_flat: false` we could return the shart when possible and otherwise return the base/flat note.

I'm also wondering why I made the sharp the first part and the flat the last 🤔   To me flat is related to less / below, so I would currently think of putting it on the left.  For different examples:

`['Cb', 'D#']` instead of `['C#', 'Db']`

`['Fb', 'E']` instead of `['E', 'Fb']`

`['F', 'E#']` instead of `['E#', 'F']`

The `.first` and `.last` calls would need to switch, of course.

If we were consistent with the rule of "flats before bases" and "bases before sharps" in the `NOTES` array, I think the above could work.

What do you think?

Reply:
Hmm I'm not sure, this is much more complex than I initially thought 🤣  - if your song is in the key of E you may have the following chord progression:
> Eb -> Cm -> Ab

With `give_flat: false` you would get:
> D# -> Cm -> G#

With your suggested solution and `give_flat: true` you would get:
> Eb -> B#m -> Ab

When in fact you wouldn't want the Cm to change because in the key of Eb the correct spelling is Cm rather than B#m.

This due to the fact that there are 2 special cases for enharmonics. There is only one semitone between E -> F and B -> C. Whereas every other note there are 2 semitones between each whole letter.

I don't really know where to go from here, it feels that to do this 100% accurately you need to know the context of what key your note is in. That feels outside the scope of this PR and maybe even outside the scope of the gem! In a lot of cases, the current solution works perfectly.

Reply:
Sorry for not responding sooner!

Yeah, I see what you mean.  That makes perfect sense 🤔 

In the spirit of moving forward I'll merge and release a new version.  A new PR is welcome, though we could keep discussing here, as you like.

It seems like the goal is maybe to prefer a non flat/sharp version (`Cm` over `B#m`).  So maybe if we had a `give_flat: true` option that's what it would do.  Or it could be `give_augmented` or something (?)